techmatters/hrm
# Set up Resource DB project

..so a healthcheck endpoint that verifies connectivity to the resources schema

## Description
This PR:
- Adds a new module as an npm workspace, `resources-service`. This service has:
  - A script to run migrations, using a connection with user `resources`.
  :exclamation: This user must exist beforehand. Since `hrm` is the master user of `hrmdb` DB, the new user must be created with `hrm` user (see below, under the changes in `hrm-service`)
  -  A migration which creates a schema `resources`, for which the new `resources` user is the owner (and only one with access to it).
  - A configuration file to interact with the DB as `resources` user.
  - A `stub` public function, which just initiates a connection to the DB using above, and runs the query `SELECT 1;`.
- Several changes in `hrm-service`:
  - A new migration is added, which creates the `resources` user and grants the necessary permissions to it in order to allow it create it's own schema within `hrmdb`.
  :exclamation: This migration requires an environment variable `RESOURCES_PASSWORD`, which will be used as the password for the newly created user. This will in turn be needed for the database connections of this user in `resources-service`. I think we'll want to provide this credentials into SSM from terraform, and bundle them in the environment variables of our `hrm-service` containers in the way we've been doing so. I just mention this, as it's a pre-requisite for this changes.
  - Adds the newly created `resources-service` package as a dependency.
  - Exposes the `stub` method from `resources-service` under the http server. Can be accessed by performing a `GET` on `/resources`.
  - Several changes in the containerized DBs and Dockerfiles to mimic the actual RDS Postgres instances. In those, `hrm` is not the superuser, it's just a user created by the RDS superuser, to which then a lot of privileges are granted. This makes easier to test that the migrations will actually behave the same in the instances as in the containerized DBs.
- Adds a build step for `hrm-ci`, since `hrm-service` now depends on a local package that must be compiled in order for the tests to be able to use it.
- Changes in `hrm-service/Dockerfile` and `hrm-service/docker-entrypoint`, the files that describe how the `hrm-service` container is actually built. :bangbang: I am not sure of this changes, and I don't know of a way to test this other than actually deploying. Ideas? :p

I'm a bit reluctant to continue changing the different npm scripts in `hrm-service` and root modules, but a reorg will soon required.

### Checklist
- [x] [Corresponding issue has been opened](https://tech-matters.atlassian.net/jira/software/c/projects/CHI/boards/7?modal=detail&selectedIssue=CHI-1137)
- [x] New tests added (not many but some :sweat_smile:)

### Verification steps
- `npm ci && npm run build`.
- Start the container DB
  - `npm run docker:compose:db:up` to start the "docker database" or,
  - `npm run docker:compose:test:up -w=hrm-service` to start the "test database".
- Run the migrations for both packages
  `npm run test:service:migrate`
  This command will run first the migrations from `hrm-service`, then the ones for `resources-service`. Note that the later won't contain a `RESOURCES_PASSWORD`, but that's fine cause the local container DBs have "trust" connection. This will be an issue if we don't provide the correct env var for actual deployments.
- At this point, the migrations should have succeed, the new user and schemas should exist. You can confirm using pgAdming (or psql). If you connect as `hrm` user, you should not be able to create/modify/do anything on `resources` schema. If you connect as `resources` user, you should not be able to create/modify/do anything on `public` schema, except on `public.SequelizeMeta` meta (this is required to run migrations using `resources-service` connection).
- Start the `hrm-service` http server
  `npm run start:service`
- After all of this, if you run
  `❯ curl -X GET 'http://localhost:3000/v0/accounts/ACd8a2e89748318adf6ddff7df6948deaf/resources' -H "Authorization: Bearer <valid flex token for development account>` 
  you should receive a response like `{"?column?":1}`, which means that the stub endpoint in `resources-service` is able to connect to the DB (hence run the query). If you stop the DB and try again, you'll get an error 500.
